### PR TITLE
Add gapply and unit tests

### PR DESCRIPTION
**Pull request by vlad17**
Gapply is in its own file, with a warning that it's going to move to spark soon. Doctests tested manually.

Changes to non-gapply things:
1. Perform spark setup in `@fixtureReuseSparkSession` BEFORE calling `setUpClass()` to let classes have access to a spark session.
2. Add `*~` pattern to gitignore. I couldn't deal with cleaning untracked files myself anymore.

### CODE REVIEW

File: python/spark_sklearn/test_utils.py [hunk 1]
```diff
@@ -23,20 +23,23 @@
...
+#
+# The SparkSession is set up before invoking the class' own set up and torn down after the
+# class' tear down, so you may safer to it in those methods.
 def fixtureReuseSparkSession(cls):
     setup = getattr(cls, 'setUpClass', None)
     teardown = getattr(cls, 'tearDownClass', None)
     def setUpClass(cls):
-        if setup: setup()
         cls.spark = SparkSession.builder.master("local").appName("Unit Tests").getOrCreate()
+        if setup: setup()
```
Comment (thunterdb) [inline python/spark_sklearn/test_utils.py]: next line (spark style guide)

File: python/spark_sklearn/tests/test_gapply.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
+
+from itertools import chain
```
Comment (thunterdb) [inline python/spark_sklearn/tests/test_gapply.py]: sort the imports

Comment (vlad17) [inline python/spark_sklearn/tests/test_gapply.py]: done

File: python/spark_sklearn/test_utils.py (hunk 1 above)
Comment (vlad17) [inline python/spark_sklearn/test_utils.py]: done

File: python/spark_sklearn/tests/test_gapply.py (hunk 2 above)
Comment (vlad17) [inline python/spark_sklearn/tests/test_gapply.py]: actually didn't change anything - already sorted, no?

Comment (thunterdb) [inline python/spark_sklearn/tests/test_gapply.py]: Oh yes, they are now.

Comment (thunterdb) [inline python/spark_sklearn/tests/test_gapply.py]: Unused import actually

File: python/spark_sklearn/tests/test_gapply.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+import pandas as pd
+import random
+import unittest
+
+from pyspark.conf import SparkConf
```
Comment (thunterdb) [inline python/spark_sklearn/tests/test_gapply.py]: unused

File: python/spark_sklearn/tests/test_gapply.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+from pyspark.sql import SparkSession
+from pyspark.sql.types import *
+from pyspark.sql import Row
```
Comment (thunterdb) [inline python/spark_sklearn/tests/test_gapply.py]: unused

File: python/spark_sklearn/tests/test_gapply.py (hunk 3 above)
Comment (vlad17) [inline python/spark_sklearn/tests/test_gapply.py]: done

File: python/spark_sklearn/tests/test_gapply.py (hunk 2 above)
Comment (vlad17) [inline python/spark_sklearn/tests/test_gapply.py]: done

File: python/spark_sklearn/tests/test_gapply.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+from pyspark.sql.tests import PythonOnlyPoint, PythonOnlyUDT, ExamplePoint, ExamplePointUDT
+
+from spark_sklearn.test_utils import fixtureReuseSparkSession
+from spark_sklearn.group_apply import gapply
+
+def _assert_frame_equal(actual, expected):
```
Comment (thunterdb) [inline python/spark_sklearn/tests/test_gapply.py]: rename to __assert_pdframe_equal so that it is clear which dataframe we are talking about

File: python/spark_sklearn/tests/test_gapply.py (hunk 4 above)
Comment (vlad17) [inline python/spark_sklearn/tests/test_gapply.py]: done

File: python/spark_sklearn/tests/test_gapply.py (hunk 5 above)
Comment (vlad17) [inline python/spark_sklearn/tests/test_gapply.py]: done

File: python/spark_sklearn/tests/test_gapply.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    # Points are unhashable, so pandas' assert_frame_equal can't check they're the same by default.
+    # We need to convert them to something that can be checked.
+    def convert_points(pt):
+        if any(isinstance(pt, t) for t in (PythonOnlyPoint, ExamplePoint)):
+            return (pt.x, pt.y)
+        return pt
+    def normalize(df):
+        converted = df.apply(lambda col: col.apply(convert_points))
```
Comment (thunterdb) [inline python/spark_sklearn/tests/test_gapply.py]: can you use a list comprehension instead?

Comment (thunterdb) [inline python/spark_sklearn/tests/test_gapply.py]: Since this is a dataframe, this is not going to work indeed. Can you put a comment for future readers?

File: python/spark_sklearn/tests/test_gapply.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        ordered = converted.sort_values(df.columns.tolist())
+        unindexed = ordered.reset_index(drop=True)
```
Comment (thunterdb) [inline python/spark_sklearn/tests/test_gapply.py]: what does it do?

File: python/spark_sklearn/tests/test_gapply.py [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+        return unindexed
+    actual, expected = (normalize(x) for x in (actual, expected))
```
Comment (thunterdb) [inline python/spark_sklearn/tests/test_gapply.py]: you do not need a sequence comprehension, do you? Make it a list comprehension

Comment (thunterdb) [inline python/spark_sklearn/tests/test_gapply.py]: Also, if this is for 2 values, you can unroll it...

File: python/spark_sklearn/tests/test_gapply.py [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+    pd.util.testing.assert_frame_equal(actual, expected)
+
+def _emptyFunc(key, vals): return pd.DataFrame.from_records([])
```
Comment (thunterdb) [inline python/spark_sklearn/tests/test_gapply.py]: 2 lines please.

File: python/spark_sklearn/test_utils.py [hunk 10]
```diff
@@ -23,20 +21,25 @@
 # unittest methods setUpClass() and tearDownClass(), invoked by the nosetest command before
 # and after unit tests are run. This enables us to create one PySpark SparkSession per
 # test fixture. The session can be referred to with self.spark or ClassName.spark.
+#
+# The SparkSession is set up before invoking the class' own set up and torn down after the
+# class' tear down, so you may safer to it in those methods.
```
Comment (thunterdb) [inline python/spark_sklearn/test_utils.py]: ??? grammar

File: python/spark_sklearn/group_apply.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+from pyspark import SparkContext
+from pyspark.sql import DataFrame, SparkSession
+from pyspark.sql.functions import explode, udf
+from pyspark.sql.types import *
+from pyspark.sql.types import Row
+
+
+def gapply(grouped_data, func, schema, *cols):
+    """WARNING: this function is planned to be migrated to apache/spark for release 2.1. This
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: remove the warning. In any case, it will become a wrapper around the spark implementation for compat.

File: python/spark_sklearn/tests/test_gapply.py (hunk 6 above)
Comment (vlad17) [inline python/spark_sklearn/tests/test_gapply.py]: I can't think of a comment that wouldn't be redundant with the code (bc variable names already indicate it) - what do you want me to say?

File: python/spark_sklearn/tests/test_gapply.py (hunk 7 above)
Comment (vlad17) [inline python/spark_sklearn/tests/test_gapply.py]: pandas saves the permutation after sorting, so assert_frame_equal would fire if the order was diff. I'll add a comment

File: python/spark_sklearn/group_apply.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    API will not be maintained.
+
+    Applies the function `func` to the grouped data; in particular, by default this calls
+    `func(key1, key2, ..., keyn, values)` where the number and order of the key arguments is
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: I suggest a slight change: make the keys a tuple instead. This way, the function does not need to know the actual structure of the key when being written.

File: python/spark_sklearn/tests/test_gapply.py (hunk 8 above)
Comment (vlad17) [inline python/spark_sklearn/tests/test_gapply.py]: but what about all that code duplication? ;)

done

File: python/spark_sklearn/group_apply.py [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+    determined by columns on which this instance's parent :class:`DataFrame` was grouped and
+    `values` is a `pandas.DataFrame` of columns selected by `cols`, in that order.
+
+    `func` is expected to return a `pandas.DataFrame` of the specified schema `schema`,
+    which should be of type :class:`StructType` (output columns are of this name and order).
+
+    If `spark.conf.get("spark.sql.retainGroupColumns")` is not `u'true'`, then `func` is
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: since config are immutable after initialization, it means you cannot mix up with keys and without keys.
This is why I suggest the function always accept the keys, but as as single tuple object or a single object so that it is easy to ignore.

File: python/spark_sklearn/tests/test_gapply.py (hunk 9 above)
Comment (vlad17) [inline python/spark_sklearn/tests/test_gapply.py]: done

File: python/spark_sklearn/group_apply.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
+"""
+Method for applying arbitrary UDFs over grouped data.
+"""
+
+from distutils.version import LooseVersion
+from itertools import chain
+from py4j import java_gateway
+
+from pyspark import SparkContext
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: unused

File: python/spark_sklearn/test_utils.py (hunk 10 above)
Comment (vlad17) [inline python/spark_sklearn/test_utils.py]: Are you referring to the dropped s for singular possessive ending in s? That's ok.

File: python/spark_sklearn/group_apply.py (hunk 11 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: done

File: python/spark_sklearn/group_apply.py [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+from pyspark.sql.functions import explode, udf
+from pyspark.sql.types import *
+from pyspark.sql.types import Row
+
+
+def gapply(grouped_data, func, schema, *cols):
+    """WARNING: this function is planned to be migrated to apache/spark for release 2.1. This
+    API will not be maintained.
+
+    Applies the function `func` to the grouped data; in particular, by default this calls
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: You go straight into technical details, which makes it hard to understand why it is useful. How about something close to: "Applies a python function to data grouped by key. Given a Spark dataframe grouped by some key, calls the function xxx on all the rows associated to this key. The rows are grouped together in a PD DataFrame ".

File: python/spark_sklearn/group_apply.py (hunk 12 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: I considered this at first (it's actually an easier impl). However, I have to disagree with the benefits of this:
1. It's arguable that the user *should* know the actual structure of the key. This is sort of like a type-check.
2. It's really inconvenient because most cases have a single key, and unwrapping that is annoying.
3. It's possible to avoid structure if you really don't want to by defining your function as `f(*args)`, then letting `args[-1]` be values.

File: python/spark_sklearn/group_apply.py [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+
+    .. :note: Users must ensure that the grouped values for every group must fit entirely in
+    memory.
+    .. :note: This method is only available if Pandas is installed.
+
+    :param func: a two argument closure, which may be either a lambda or named function
+    :param schema: the return schema for `func`
+    :param cols: list of column names (string only)
+
+    :raise ValueError: if `*` is in `cols`
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: ?

File: python/spark_sklearn/group_apply.py (hunk 13 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: See above point - we can discuss in person if you disagree.

File: python/spark_sklearn/group_apply.py (hunk 14 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: done

File: python/spark_sklearn/group_apply.py [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+    ...                       Row(course="Java",   year=2013, earnings=30000)])
+    ...     .select("course", "year", "earnings"))
+    >>> def yearlyMedian(key, vals):
+    ...     all_years = set(vals['year'])
+    ...     # Note that interpolation is performed, so we need to cast back to long.
+    ...     yearly_median = [(key, year, long(vals['earnings'][vals['year'] == year].median()))
+    ...                      for year in all_years]
+    ...     return pd.DataFrame.from_records(yearly_median)
+    >>> gapply(df.groupBy("course"), yearlyMedian, df.schema).orderBy("earnings").show()
+    +------+----+--------+
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: I think this example would be easier to understand if you just grouped by course and computed the median, without the year.

File: python/spark_sklearn/group_apply.py (hunk 12 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: 4. Parallels the way `cols` are specified - if it was convenient there it's convenient here.

File: python/spark_sklearn/group_apply.py (hunk 17 above)
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: Then you could show how to compute the new schema

File: python/spark_sklearn/group_apply.py [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+    |dotNET|2012|    7500|
+    |  Java|2012|   20000|
+    |  Java|2013|   30000|
+    |dotNET|2013|   48000|
+    +------+----+--------+
+    <BLANKLINE>
+    >>> def twoKeyYearlyMedian(course, year, vals):
+    ...     return pd.DataFrame.from_records([(course, year, long(vals["earnings"].median()))])
+    >>> gapply(df.groupBy("course", "year"), twoKeyYearlyMedian, df.schema, "earnings") \
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: based on the comment above regarding the keys, then you would not need a special example for multiple keys.

File: python/spark_sklearn/group_apply.py [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+    |course|year|earnings|
+    +------+----+--------+
+    |dotNET|2012|    7500|
+    |  Java|2012|   20000|
+    |  Java|2013|   30000|
+    |dotNET|2013|   48000|
+    +------+----+--------+
+    <BLANKLINE>
+    """
+    import pandas
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: I think the customary import is `import pandas as pd`

File: python/spark_sklearn/group_apply.py (hunk 15 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: done

File: python/spark_sklearn/group_apply.py [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+    if LooseVersion(pandas.__version__) < LooseVersion(minPandasVersion):
+        raise ImportError('Pandas installed but version is {}, {} required'
+                          .format(pandas.__version__, minPandasVersion))
+
+    # Do a null aggregation to retrieve the keys first (should be no computation)
+    key_cols = grouped_data.agg({}).columns
+
+    if not cols:
+        # Extract the full column list with the parent df
+        javaDFName = "org$apache$spark$sql$RelationalGroupedDataset$$df"
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: yeah, it's not pretty...

File: python/spark_sklearn/group_apply.py (hunk 16 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: If you try to use the asterisk for column expansion, it would throw a weird SQL error by default. I'm saying you can't do that early here.

File: python/spark_sklearn/group_apply.py [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+
+    The order of the rows passed in as Pandas rows is not guaranteed to be stable relative to 
+    the original row order.
+
+    .. :note: Users must ensure that the grouped values for every group must fit entirely in
+    memory.
+    .. :note: This method is only available if Pandas is installed.
+
+    :param func: a two argument closure, which may be either a lambda or named function
+    :param schema: the return schema for `func`
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: should be a sql structtype

File: python/spark_sklearn/group_apply.py (hunk 17 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: Ok. I'll keep the twokey example, though, since that shows something else.

File: python/spark_sklearn/group_apply.py [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+    if "*" in cols:
+        raise ValueError("cols expected to contain only singular columns")
+
+    if len(set(cols)) < len(cols):
+        raise ValueError("cols expected not to contain duplicate columns")
+
+    if not isinstance(schema, StructType):
+        raise ValueError("output schema should be a StructType")
+
+    inputAggDF = grouped_data.agg({col: 'collect_list' for col in cols})
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: no need for `[]`?

File: python/spark_sklearn/group_apply.py [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+    # Recover canonical order (aggregation may change column order)
+    cannonicalOrder = chain(key_cols, (inputAggDF['collect_list(' + col + ')'] for col in cols))
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: list comp please

File: python/spark_sklearn/group_apply.py (hunk 22 above)
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: I was not familiar with this syntax. Can you use the explicit `dict` constructor?

File: python/spark_sklearn/group_apply.py [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+    inputAggDF = inputAggDF.select(*cannonicalOrder)
+
+    def pandasWrappedFunc(*args):
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: put a comment to explain what it does

File: python/spark_sklearn/group_apply.py [hunk 25]
```diff
@@ -0,0 +1,158 @@
...
+        nvals = len(cols)
+        keys, collectedCols = args[:-nvals], args[-nvals:]
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: 2 lines

File: python/spark_sklearn/group_apply.py [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+        valuesDF = pandas.DataFrame.from_dict(
+            {colName: colList for colName, colList in zip(cols, collectedCols)})
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: `dict(zip(...))`

File: python/spark_sklearn/group_apply.py (hunk 12 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: (after talking - done).

File: python/spark_sklearn/group_apply.py (hunk 13 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: after offline discussion, done

File: python/spark_sklearn/group_apply.py (hunk 18 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: (as discussed, still kept this)

File: python/spark_sklearn/group_apply.py (hunk 19 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: I think DataFrame.toPandas() doesn't do this but done anyway.

File: python/spark_sklearn/group_apply.py (hunk 21 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: done

File: python/spark_sklearn/group_apply.py (hunk 22 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: Why? What's wrong with a dictionary comprehension?

File: python/spark_sklearn/group_apply.py (hunk 23 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: Why? it'll be converted to an iterator immediately anyway

File: python/spark_sklearn/group_apply.py (hunk 24 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: done

File: python/spark_sklearn/group_apply.py (hunk 25 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: In general I'd agree, but here it should be noted I'm splitting the list - is this not clearer?

File: python/spark_sklearn/group_apply.py (hunk 26 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: again, why not dict comp?

File: python/spark_sklearn/group_apply.py (hunk 20 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: i don't disagree

File: python/spark_sklearn/group_apply.py (hunk 22 above)
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: It is older than PEP-8. Let me check with authoritative people.

File: python/spark_sklearn/group_apply.py [hunk 27]
```diff
@@ -0,0 +1,175 @@
...
+    inputAggDF = inputAggDF.select(*cannonicalOrder)
+
+    # Wraps the user-proveded function with another python function, which prepares the
+    # input in the form specified by the documentation. Then, once the function completes,
+    # this wrapper prepends the keys to the output values and converts from pandas.
+    def pandasWrappedFunc(*args):
+        nvals = len(cols)
+        keys, collectedCols = args[:-nvals], args[-nvals:]
+        paramKeys = tuple(keys)
+        if len(paramKeys) == 1: paramKeys = paramKeys[0]
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: no one-liner please

Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: or use if / else

File: python/spark_sklearn/group_apply.py [hunk 28]
```diff
@@ -0,0 +1,175 @@
...
+        valuesDF = pd.DataFrame.from_dict(
+            {colName: colList for colName, colList in zip(cols, collectedCols)})
+        valuesDF = valuesDF[list(cols)] # reorder to cannonical
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: canonical

File: python/spark_sklearn/group_apply.py [hunk 29]
```diff
@@ -0,0 +1,175 @@
...
+
+    # Wraps the user-proveded function with another python function, which prepares the
+    # input in the form specified by the documentation. Then, once the function completes,
+    # this wrapper prepends the keys to the output values and converts from pandas.
+    def pandasWrappedFunc(*args):
+        nvals = len(cols)
+        keys, collectedCols = args[:-nvals], args[-nvals:]
+        paramKeys = tuple(keys)
+        if len(paramKeys) == 1: paramKeys = paramKeys[0]
+        valuesDF = pd.DataFrame.from_dict(
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: dict(zip(...))

File: python/spark_sklearn/group_apply.py [hunk 30]
```diff
@@ -0,0 +1,175 @@
...
+    # Recover canonical order (aggregation may change column order)
+    cannonicalOrder = chain(keyCols, (inputAggDF['collect_list(' + col + ')'] for col in cols))
+    inputAggDF = inputAggDF.select(*cannonicalOrder)
+
+    # Wraps the user-proveded function with another python function, which prepares the
+    # input in the form specified by the documentation. Then, once the function completes,
+    # this wrapper prepends the keys to the output values and converts from pandas.
+    def pandasWrappedFunc(*args):
+        nvals = len(cols)
+        keys, collectedCols = args[:-nvals], args[-nvals:]
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: 2 lines please

File: python/spark_sklearn/group_apply.py (hunk 27 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: done

File: python/spark_sklearn/group_apply.py (hunk 28 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: done

File: python/spark_sklearn/group_apply.py (hunk 29 above)
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: It will help the reader if you use something like `PDF` as a suffix for pandas dataframes, to clearly separate them from the spark dataframes.

Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: again, why?

Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: done

File: python/spark_sklearn/group_apply.py (hunk 30 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: done

File: python/spark_sklearn/test_utils.py (hunk 10 above)
Comment (thunterdb) [inline python/spark_sklearn/test_utils.py]: `you may safer to it`

File: python/spark_sklearn/tests/test_gapply.py (hunk 6 above)
Comment (thunterdb) [inline python/spark_sklearn/tests/test_gapply.py]: I see. Nevermind.

File: python/spark_sklearn/tests/test_gapply.py [hunk 31]
```diff
@@ -0,0 +1,216 @@
...
+        self.assertRaises(ValueError, gapply, gd, _emptyFunc, longLongLongSchema, "b", "c", "b")
+
+    NROWS = 15
+    NKEYS = 5
+    NVALS = 100
+
+    def checkGapplyEquivalentToPandas(self, pandasAggFunction, dataType, dataGen):
+        schema = StructType().add("val", dataType)
+        pandasDF = pd.DataFrame.from_dict({
+            "key": [random.randrange(GapplyTests.NKEYS) for _ in range(GapplyTests.NROWS)],
```
Comment (thunterdb) [inline python/spark_sklearn/tests/test_gapply.py]: initialize the generator with a constant seed before using it

File: python/spark_sklearn/group_apply.py [hunk 32]
```diff
@@ -0,0 +1,175 @@
...
+    The order of the rows passed in as Pandas rows is not guaranteed to be stable relative to 
+    the original row order.
+
+    .. :note: Users must ensure that the grouped values for every group must fit entirely in
+    memory.
+    .. :note: This method is only available if Pandas is installed.
+
+    :param func: a two argument function, which may be either a lambda or named function
+    :param schema: the return schema for `func`, a :class:`StructType`
+    :param cols: list of column names (string only)
```
Comment (thunterdb) [inline python/spark_sklearn/group_apply.py]: do you mention that the other columns are dropped? It is obvious from the implementation, but it should be mentioned somewhere.

File: python/spark_sklearn/tests/test_gapply.py (hunk 31 above)
Comment (vlad17) [inline python/spark_sklearn/tests/test_gapply.py]: Can I do this with a different PR? I was going to make a class for this with a nice random interface (by default a time-based seed is chosen, but seed is printed and can be set by an environment variable for replicating broken tests)

File: python/spark_sklearn/group_apply.py (hunk 32 above)
Comment (vlad17) [inline python/spark_sklearn/group_apply.py]: Yes -it's mentioned here:

"`values` is a `pandas.DataFrame` of columns selected by `cols`, in that order."
...
"the schema being the original key schema
    prepended to `schema`"

File: python/spark_sklearn/tests/test_gapply.py [hunk 33]
```diff
@@ -0,0 +1,216 @@
...
+        _assert_pdframe_equal(actual, expected)
+        def func(key, vals):
+            assert vals.columns.tolist() == ["val2", "val1"], vals.columns
+            return pd.DataFrame.from_records([(vals["val2"].sum(),)])
+        gd = df.select("val2", "key", "val1").groupBy("key")
+        actual = gapply(gd, func, schema).toPandas()
+        _assert_pdframe_equal(actual, expected)
+
+@fixtureReuseSparkSession
+class GapplyConfTests(unittest.TestCase):
```
Comment (thunterdb) [inline python/spark_sklearn/tests/test_gapply.py]: where do you use this class?

Comment (vlad17) [inline python/spark_sklearn/tests/test_gapply.py]: unittest uses it

File: python/spark_sklearn/tests/test_gapply.py (hunk 31 above)
Comment (thunterdb) [inline python/spark_sklearn/tests/test_gapply.py]: No need to do that. We tend to fix a seed in the beforeAll and we do not bother with time-based.

Comment (vlad17) [inline python/spark_sklearn/tests/test_gapply.py]: (alternative class scenario discussed offline)

**Comment by thunterdb** [comment]
It looks great!
